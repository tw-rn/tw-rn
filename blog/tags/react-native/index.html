<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="generator" content="Docusaurus v2.0.0-alpha.58">
<link rel="search" type="application/opensearchdescription+xml" title="tw-rn" href="/tw-rn/opensearch.xml">
<script src="https://snack.expo.io/embed.js"></script><title data-react-helmet="true">Posts tagged &quot;react native&quot; | tw-rn</title><meta data-react-helmet="true" property="og:title" content="Posts tagged &quot;react native&quot; | tw-rn"><meta data-react-helmet="true" name="description" content="Blog | Tagged &quot;react native&quot;"><meta data-react-helmet="true" property="og:description" content="Blog | Tagged &quot;react native&quot;"><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><link data-react-helmet="true" rel="shortcut icon" href="/tw-rn/img/logo.svg"><link rel="stylesheet" href="/tw-rn/styles.8ec757e3.css">
<link rel="preload" href="/tw-rn/styles.627a782c.js" as="script">
<link rel="preload" href="/tw-rn/runtime~main.912ae9ce.js" as="script">
<link rel="preload" href="/tw-rn/main.9ac002ca.js" as="script">
<link rel="preload" href="/tw-rn/1.911ba83d.js" as="script">
<link rel="preload" href="/tw-rn/2.18eece6c.js" as="script">
<link rel="preload" href="/tw-rn/3.290168cc.js" as="script">
<link rel="preload" href="/tw-rn/6875c492.d06c4903.js" as="script">
<link rel="preload" href="/tw-rn/af172acd.0c5fde42.js" as="script">
<link rel="preload" href="/tw-rn/f791411d.4431bb28.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=window.matchMedia("(prefers-color-scheme: dark)"),n=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();null!==n?t(n):e.matches&&t("dark")}()</script><div id="__docusaurus">
<nav class="navbar navbar--light navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><div aria-label="Navigation bar toggle" class="navbar__toggle" role="button" tabindex="0"><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></div><a class="navbar__brand" href="/tw-rn/"><img class="navbar__logo" src="/tw-rn/img/logo.svg" alt="tw-rn Logo"><strong class="navbar__title"></strong></a><a class="navbar__item navbar__link" href="/tw-rn/docs/fundamentals/getting-started">Docs</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/tw-rn/tw-rn" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub</a><div class="react-toggle react-toggle--disabled displayOnlyInLargeViewport_1gtM"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_keGJ moon_1gwN"></span></div><div class="react-toggle-track-x"><span class="toggle_keGJ sun_3CPA"></span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" disabled="" aria-label="Dark mode toggle" class="react-toggle-screenreader-only"></div><div class="navbar__search"><div class="searchWrapper_-F3-"><span aria-label="expand searchbar" role="button" class="searchIconButton_1Dnz" tabindex="0"></span><input type="search" id="search_input_react" placeholder="Search" aria-label="Search" class="navbar__search-input searchInput_2eto"></div></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/tw-rn/"><img class="navbar__logo" src="/tw-rn/img/logo.svg" alt="tw-rn Logo"><strong class="navbar__title"></strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" href="/tw-rn/docs/fundamentals/getting-started">Docs</a></li><li class="menu__list-item"><a href="https://github.com/tw-rn/tw-rn" target="_blank" rel="noopener noreferrer" class="menu__link">GitHub</a></li></ul></div></div></div></nav><div class="main-wrapper"><div class="container margin-vert--lg"><div class="row"><main class="col col--8 col--offset-2"><h1>1 post tagged with &quot;react native&quot;</h1><a href="/tw-rn/blog/tags">View All Tags</a><div class="margin-vert--xl"><article class="margin-bottom--xl"><header><h2 class="margin-bottom--sm blogPostTitle_2RZH"><a href="/tw-rn/blog/hello-world">Hello World</a></h2><div class="margin-vert--md"><time datetime="2019-05-30T00:00:00.000Z" class="blogPostDate_3tRe">May 30, 2019  · 1 min read</time></div><div class="avatar margin-vert--md"><a class="avatar__photo-link avatar__photo" href="https://github.com/ericktamayo" target="_blank" rel="noreferrer noopener"><img src="https://avatars3.githubusercontent.com/u/4788817?s=400&amp;u=f1fa98b8e2d26e5b8ae8ce9817fc559ec5509439&amp;v=4" alt="Erick Tamayo"></a><div class="avatar__intro"><h4 class="avatar__name"><a href="https://github.com/ericktamayo" target="_blank" rel="noreferrer noopener">Erick Tamayo</a></h4><small class="avatar__subtitle">Creator of tw-rn</small></div></div></header><section class="markdown"><p>This is the start of tw-rn</p></section><footer class="row margin-vert--lg"><div class="col"><strong>Tags:</strong><a class="margin-horiz--sm" href="/tw-rn/blog/tags/tw-rn">tw-rn</a><a class="margin-horiz--sm" href="/tw-rn/blog/tags/tailwindcss">tailwindcss</a><a class="margin-horiz--sm" href="/tw-rn/blog/tags/react-native">react native</a></div><div class="col text--right"><a aria-label="Read more about Hello World" href="/tw-rn/blog/hello-world"><strong>Read More</strong></a></div></footer></article></div></main></div></div></div><footer class="footer footer--dark"><div class="container"><div class="text--center"><div>Copyright © 2020 tw-rn</div></div></div></footer></div>
<script src="/tw-rn/styles.627a782c.js"></script>
<script src="/tw-rn/runtime~main.912ae9ce.js"></script>
<script src="/tw-rn/main.9ac002ca.js"></script>
<script src="/tw-rn/1.911ba83d.js"></script>
<script src="/tw-rn/2.18eece6c.js"></script>
<script src="/tw-rn/3.290168cc.js"></script>
<script src="/tw-rn/6875c492.d06c4903.js"></script>
<script src="/tw-rn/af172acd.0c5fde42.js"></script>
<script src="/tw-rn/f791411d.4431bb28.js"></script>
</body>
</html>