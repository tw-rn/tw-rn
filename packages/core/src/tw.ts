import merge from "deepmerge";
import memoize from "fast-memoize";
import { Variant, TailwindStyle } from "./types";

const styles = global.__TAILWINDCSS_NATIVE_STYLES__;
const styleEntries = Object.entries(styles || []);
const variantRegex = new RegExp(`^(${Object.values(Variant).join("|")}):(.*)$`);

export const generateTailwindStyle = (
  stylesArray: TemplateStringsArray,
  ...variables: string[]
): TailwindStyle => {
  if (typeof styles === "undefined") {
    __DEV__ && console.warn(`Can't find styles. Please include your CSS in your App entry point.`);
    return { media: { "": {} } };
  }

  const computedStyles = stylesArray
    .map((chunk, index) => `${chunk}${variables[index] || ""}`)
    .join("")
    .replace(/\s{2,}/g, " ")
    .split(" ")
    .filter(Boolean);

  const initial: TailwindStyle = {
    media: {},
    // hover: [],
    // focus: [],
  };

  const tailwindStyle = computedStyles.reduce<TailwindStyle>(
    (acc, computedStyle) => {
      const [, variant, variantStyleName] = variantRegex.exec(computedStyle) || [];

      // Check if is a variant
      if (variant) {
      }

      const media = styleEntries.reduce((acc, [media, styles]) => {
        const name = computedStyle.replace(":", "\\:");
        const style = styles[name];

        if (!style) {
          return acc;
        }

        return merge(acc, { [media]: style });
      }, {});

      return merge(acc, { media });
    },
    { media: { "": {} } }
  );

  // console.log({ tailwindStyle });

  // sm: -> get the
  // const mediaStyles = screens.reduce<Partial<TailwindStyle>>((acc, screen) => {

  //   return acc;
  // }, {});

  // const twStyles = {
  //   media: {
  //     // sm: {...}
  //   },
  //   hover: {},
  //   focus: {},
  // };

  // const tailwindNativeStyles = computedStyles.reduce((acc, computedStyle) => {
  //   return acc;
  // }, {});
  // const tailwindNativeStyles = computedStyles.reduce((acc, computedStyle) => {
  //   // Check if is has a variant
  //   const [, variant, styleName] = /^(\w+):(.*)/.exec(computedStyle) || [];

  //   // No variant found
  //   if (!variant) {
  //     const { mediaQuery, stylesheet } = styles[""];

  //     if (stylesheet[computedStyle]) {
  //       return merge(acc, { [""]: { mediaQuery, styles: [stylesheet[computedStyle]] } });
  //     } else {
  // __DEV__ && console.warn(`Invalid style ${computedStyle}`);
  //       return acc;
  //     }
  //   }

  //   return acc;

  // }, {});

  // const mappedStyle = styles[computedStyle];

  // if (!mappedStyle) {
  //   // Check if is a valid variant
  //   const [, variant, styleName] = variantRegex.exec(computedStyle) || [];

  //   if (!variant) {
  // __DEV__ && console.warn(`Invalid style or variant ${computedStyle}`);
  // return acc;
  //   }

  //   if (typeof styles[styleName] === "undefined") {
  //     __DEV__ && console.warn(`Style not found: ${styleName}`);
  //     return acc;
  //   }

  //   const variantStyle = styles[styleName];

  //   return [...acc, [variant, variantStyle]];
  // }

  // return [...acc, mappedStyle];

  // console.log({ tailwindNativeStyles });
  // return tailwindNativeStyles;
  // console.log({ tailwindStyle });

  return tailwindStyle;
};

// const serializer = (...arg: any) => {
//   console.log({ arg });
//   return "1";
// };

export const tw = memoize(generateTailwindStyle, { strategy: memoize.strategies.variadic });
