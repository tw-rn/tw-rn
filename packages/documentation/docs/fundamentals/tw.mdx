---
id: tw
title: tw
---

The `tw` function parses the TailwindCSS classes string to its corresponding React Native styles. These styles are classified depending on the selector that are in the string. This function is used by the `withTwrn` HoC to be able to apply the correct styles to wrapped components.

```js
// const styles = { __: { "": { backgrounColor: "#ffffff" } } }
const styles = tw`bg-white`;
```
