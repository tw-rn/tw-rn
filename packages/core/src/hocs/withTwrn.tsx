import React, { ComponentType, useMemo, useCallback, useRef, useEffect, useState } from "react";
import { NativeSyntheticEvent, TargetedEvent, StyleProp, Animated } from "react-native";
import { TailwindReactNativeStyle } from "../types";
import {
  useTailwindReactNativeStyle,
  usePlatformStyles,
  useMediaStyles,
  useHoverStyles,
  useFocusStyles,
  useOrientationStyles,
  useAnimationStyles,
  useCombineStyles,
} from "../hooks";

export type VariantProps<P, O extends keyof P> = {
  onMouseEnter?: (e: any) => void;
  onMouseLeave?: (e: any) => void;
  onFocus?: (e: NativeSyntheticEvent<TargetedEvent>) => void;
  onBlur?: (e: NativeSyntheticEvent<TargetedEvent>) => void;
} & { [key in O]?: StyleProp<TailwindReactNativeStyle & P[O]> } &
  Omit<P, O>;

export const withTwrn = <P extends object, O extends keyof P>(
  Component: ComponentType<P>,
  styleKeys: O[]
): ComponentType<VariantProps<P, O>> => ({
  onMouseEnter,
  onMouseLeave,
  onFocus,
  onBlur,
  ...props
}) => {
  const tailwindReactNativeStyle = useTailwindReactNativeStyle(props, styleKeys);

  const platformStyles = usePlatformStyles(tailwindReactNativeStyle);

  const mediaStyles = useMediaStyles(platformStyles);

  const { hoverStyles, handleOnMouseEnter, handleOnMouseLeave } = useHoverStyles(
    platformStyles,
    onMouseEnter,
    onMouseLeave
  );

  const { focusStyles, handleOnFocus, handleOnBlur } = useFocusStyles(
    platformStyles,
    onFocus,
    onBlur
  );

  const orientationStyles = useOrientationStyles(platformStyles);

  const combinedStyles = useCombineStyles([
    mediaStyles,
    orientationStyles,
    hoverStyles,
    focusStyles,
  ]);

  const { needsAnimatedComponent, regularOrAnimatedStyles } = useAnimationStyles(
    platformStyles,
    combinedStyles
  );

  // Combining styles position with keys indicated in the styleKeys
  const regularOrAnimatedStylesProps = useMemo(
    () =>
      styleKeys.reduce(
        (acc, key, index) => ({ ...acc, [key]: regularOrAnimatedStyles[index] }),
        {}
      ),
    [regularOrAnimatedStyles]
  );

  // If combined style is null, it means that we're in SSR and should not
  // render because we don't have the destination size. Note: can be improved.
  if (combinedStyles.some((style) => style === null)) return null;

  const ComponentToRender = useMemo(() => {
    needsAnimatedComponent ? () => null : Component;
  }, [needsAnimatedComponent])

  return needsAnimatedComponent ? (
    <Component
      {...(props as P)}
      {...regularOrAnimatedStylesProps}
      onMouseEnter={handleOnMouseEnter}
      onMouseLeave={handleOnMouseLeave}
      onFocus={handleOnFocus}
      onBlur={handleOnBlur}
    />
  ) : (
    <Component
      {...(props as P)}
      {...regularOrAnimatedStylesProps}
      onMouseEnter={handleOnMouseEnter}
      onMouseLeave={handleOnMouseLeave}
      onFocus={handleOnFocus}
      onBlur={handleOnBlur}
    />
  );
};

export default withTwrn;
