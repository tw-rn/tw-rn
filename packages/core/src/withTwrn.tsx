import React, { useMemo, useCallback, useState, FunctionComponent, ComponentType } from "react";
import { StyleSheet, Platform, NativeSyntheticEvent, TargetedEvent } from "react-native";
import { Style } from "./types";
import { convertToTailwindReactNativeStyle } from "./helpers";
import { usePlatformStyle, useMediaStyle, useTailwindReactNativeStyle } from "./hooks";

export type TailwindNativeHocProps = {
  style?: Style;
  onMouseEnter?: (e: any) => void;
  onMouseLeave?: (e: any) => void;
  onFocus?: (e: NativeSyntheticEvent<TargetedEvent>) => void;
  onBlur?: (e: NativeSyntheticEvent<TargetedEvent>) => void;
};

const withTwrn = <P extends object>(
  Component: ComponentType<P>
): FunctionComponent<TailwindNativeHocProps & Omit<P, "style">> => ({
  style,
  onMouseEnter,
  onMouseLeave,
  onFocus,
  onBlur,
  ...props
}) => {
  const tailwindReactNativeStyle = useTailwindReactNativeStyle(style);

  const platformStyle = usePlatformStyle(tailwindReactNativeStyle);

  const mediaStyle = useMediaStyle(platformStyle);

  console.log({ platformStyle, mediaStyle });

  // const [isHovered, setIsHovered] = useState(false);

  // const [isFocused, setIsFocused] = useState(false);

  // const handleMouseEnter = useCallback(
  //   (e) => {
  //     setIsHovered(true);
  //     onMouseEnter && onMouseEnter(e);
  //   },
  //   [onMouseEnter]
  // );

  // const handleMouseLeave = useCallback(
  //   (e) => {
  //     setIsHovered(false);
  //     onMouseLeave && onMouseLeave(e);
  //   },
  //   [onMouseLeave]
  // );

  // const handleFocus = useCallback(
  //   (e) => {
  //     setIsFocused(true);
  //     onFocus && onFocus(e);
  //   },
  //   [onFocus]
  // );

  // const handleBlur = useCallback(
  //   (e) => {
  //     setIsFocused(false);
  //     onBlur && onBlur(e);
  //   },
  //   [onBlur]
  // );

  // console.log({ style });

  // const platformComposed = useMemo(() => {
  //   // const { native, web, ios, android } = reducedStyles;
  //   // return StyleSheet.compose(native, Platform.select({ web, ios, android }));
  // }, [reducedStyles]);

  // const composedStyle = useMemo(() => {
  //   const { hover, focus } = reducedStyles;

  //   const hoverComposed = isHovered
  //     ? StyleSheet.compose(platformComposed, hover)
  //     : platformComposed;

  //   const focusComposed = isFocused ? StyleSheet.compose(hoverComposed, focus) : hoverComposed;

  //   return focusComposed;
  // }, [platformComposed, isHovered, isFocused]);

  // const composedStyle = style
  // if (Platform.OS === "web" && typeof window === "undefined") return null;

  // if (currentMediaQueryValue === null) return null;

  // console.log({ mediaStyle });

  return (
    <Component
      style={mediaStyle}
      // style={mediaStyle}
      // onMouseEnter={handleMouseEnter} // Should work only if this has a class that uses this
      // onMouseLeave={handleMouseLeave} // Should work only if this has a class that uses this
      // onFocus={handleFocus} // Should work only if this has a class that uses this
      // onBlur={handleBlur} // Should work only if this has a class that uses this
      {...(props as P)}
    />
  );
};

export default withTwrn;
