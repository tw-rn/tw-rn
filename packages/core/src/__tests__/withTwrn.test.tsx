import React from "react";
import { View as RnView, TextInput as RnTextInput, Platform } from "react-native";
import { render, fireEvent, NativeTestEvent, wait, toJSON } from "@testing-library/react-native";
import withTwrn from "../withTwrn";
import { tw } from "../tw";

const View = withTwrn(RnView);
const TextInput = withTwrn(RnTextInput);

describe("withTwrn", () => {
  afterAll(() => {
    Platform.OS = "ios";
  });

  it("should render correctly with no styles", () => {
    const { getByTestId } = render(<View testID="view" />);
    expect(toJSON(getByTestId("view"))).toMatchSnapshot();
  });

  it("should render regular react-native styles", () => {
    const { getByTestId, rerender } = render(
      <View testID="view" style={{ backgroundColor: "#ff0000" }} />
    );

    expect(toJSON(getByTestId("view"))).toMatchSnapshot();

    rerender(<View testID="view" style={[{ backgroundColor: "#ff0000" }, { padding: 10 }]} />);

    expect(toJSON(getByTestId("view"))).toMatchSnapshot();
  });

  it("should render tw-rn base styles correctly in ios", () => {
    const { getByTestId } = render(<View testID="view" style={tw`bg-white`} />);

    expect(toJSON(getByTestId("view"))).toMatchSnapshot();
  });

  it("should render tw-rn base and regular react-native styles correctly", () => {
    const { getByTestId } = render(<View testID="view" style={[tw`bg-white`, { padding: 10 }]} />);

    expect(toJSON(getByTestId("view"))).toMatchSnapshot();
  });

  it("should render correctly the hover: pseudo selector", async () => {
    const handleOnMouseEnter = jest.fn();
    const handleOnMouseLeave = jest.fn();

    const { getByTestId, asJSON, rerender } = render(
      <View
        testID="view"
        style={tw`hover:bg-white`}
        onMouseEnter={handleOnMouseEnter}
        onMouseLeave={handleOnMouseLeave}
      />
    );

    // Idle
    expect(toJSON(getByTestId("view"))).toMatchSnapshot("hover: - idle");

    // Hover
    fireEvent(getByTestId("view"), new NativeTestEvent("mouseEnter", { nativeEvent: {} }));

    await wait(() => expect(handleOnMouseEnter).toBeCalled());

    rerender(
      <View
        testID="view"
        style={tw`hover:bg-white`}
        onMouseEnter={handleOnMouseEnter}
        onMouseLeave={handleOnMouseLeave}
      />
    );

    expect(toJSON(getByTestId("view"))).toMatchSnapshot("hover: - hovered");

    // No hover
    fireEvent(getByTestId("view"), new NativeTestEvent("mouseLeave", { nativeEvent: {} }));

    await wait(() => expect(handleOnMouseLeave).toBeCalled());

    rerender(
      <View
        testID="view"
        style={tw`hover:bg-white`}
        onMouseEnter={handleOnMouseEnter}
        onMouseLeave={handleOnMouseLeave}
      />
    );

    expect(toJSON(getByTestId("view"))).toMatchSnapshot("hover: - no-hover");
  });

  it("should render correctly the focus: pseudo selector", async () => {
    const handleOnFocus = jest.fn();
    const handleOnBlur = jest.fn();

    const { getByTestId, asJSON, rerender } = render(
      <TextInput
        testID="text-input"
        style={tw`focus:bg-white`}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
    );

    // Idle
    expect(toJSON(getByTestId("text-input"))).toMatchSnapshot("focus: - idle");

    // Focus
    fireEvent.focus(getByTestId("text-input"));

    await wait(() => expect(handleOnFocus).toBeCalled());

    rerender(
      <TextInput
        testID="text-input"
        style={tw`focus:bg-white`}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
    );

    expect(toJSON(getByTestId("text-input"))).toMatchSnapshot("focus: - focused");

    // Blur
    fireEvent.blur(getByTestId("text-input"));

    await wait(() => expect(handleOnBlur).toBeCalled());

    rerender(
      <TextInput
        testID="text-input"
        style={tw`focus:bg-white`}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      />
    );

    expect(toJSON(getByTestId("text-input"))).toMatchSnapshot("focus: - blurred");
  });

  describe("ios", () => {
    beforeAll(() => {
      Platform.OS = "ios";
    });

    it("should not render media queries", () => {
      const { getByTestId } = render(<View testID="view" style={tw`sm:bg-red`} />);

      expect(toJSON(getByTestId("view"))).toMatchSnapshot();
    });
  });

  describe("android", () => {
    beforeAll(() => {
      Platform.OS = "android";
    });

    it("should not render media queries on android", () => {
      const { asJSON } = render(<View testID="view" style={tw`sm:bg-red`} />);

      expect(asJSON()).toMatchSnapshot();
    });
  });

  describe("web", () => {
    beforeAll(() => {
      Platform.OS = "web";
    });

    afterAll(() => {
      delete global.window;
    });

    beforeEach(() => {
      (global.window as any) = {
        matchMedia: jest.fn().mockImplementation(() => ({
          matches: true,
          addListener: jest.fn(),
          removeListener: jest.fn(),
        })),
      };
    });

    it("should not render on SSR", () => {
      delete global.window;

      const { asJSON } = render(<View testID="view" style={tw`sm:bg-red`} />);

      expect(asJSON()).toMatchSnapshot();
    });

    it("should render media queries on web", () => {
      const { getByTestId } = render(<View testID="view" style={tw`sm:bg-red`} />);

      expect(toJSON(getByTestId("view"))).toMatchSnapshot();
    });

    it("should render non media queries styles along media queries styles", () => {
      const { getByTestId } = render(<View testID="view" style={tw`p-1 sm:bg-red`} />);

      expect(toJSON(getByTestId("view"))).toMatchSnapshot();
    });

    it("should respect queries order styles", () => {
      const { getByTestId } = render(<View testID="view" style={tw`p-1 sm:bg-red md:bg-green`} />);

      expect(toJSON(getByTestId("view"))).toMatchSnapshot();
    });
  });
});
