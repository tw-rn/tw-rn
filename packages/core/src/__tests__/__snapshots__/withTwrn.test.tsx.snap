// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withTwrn android should not render media queries on android 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={Object {}}
    testID="view"
  />
</View>
`;

exports[`withTwrn ios should not render media queries 1`] = `
<View
  style={Object {}}
  testID="view"
/>
`;

exports[`withTwrn should render correctly the focus: pseudo selector: focus: - blurred 1`] = `
<TextInput
  allowFontScaling={true}
  rejectResponderTermination={true}
  style={Object {}}
  testID="text-input"
  underlineColorAndroid="transparent"
/>
`;

exports[`withTwrn should render correctly the focus: pseudo selector: focus: - focused 1`] = `
<TextInput
  allowFontScaling={true}
  rejectResponderTermination={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
  testID="text-input"
  underlineColorAndroid="transparent"
/>
`;

exports[`withTwrn should render correctly the focus: pseudo selector: focus: - idle 1`] = `
<TextInput
  allowFontScaling={true}
  rejectResponderTermination={true}
  style={Object {}}
  testID="text-input"
  underlineColorAndroid="transparent"
/>
`;

exports[`withTwrn should render correctly the hover: pseudo selector: hover: - hovered 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
  testID="view"
/>
`;

exports[`withTwrn should render correctly the hover: pseudo selector: hover: - idle 1`] = `
<View
  style={Object {}}
  testID="view"
/>
`;

exports[`withTwrn should render correctly the hover: pseudo selector: hover: - no-hover 1`] = `
<View
  style={Object {}}
  testID="view"
/>
`;

exports[`withTwrn should render correctly with no styles 1`] = `
<View
  style={Object {}}
  testID="view"
/>
`;

exports[`withTwrn should render regular react-native styles 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff0000",
    }
  }
  testID="view"
/>
`;

exports[`withTwrn should render regular react-native styles 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff0000",
      "padding": 10,
    }
  }
  testID="view"
/>
`;

exports[`withTwrn should render tw-rn base and regular react-native styles correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
      "padding": 10,
    }
  }
  testID="view"
/>
`;

exports[`withTwrn should render tw-rn base styles correctly in ios 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ffffff",
    }
  }
  testID="view"
/>
`;

exports[`withTwrn web should not render on SSR 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`withTwrn web should render media queries on web 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff0000",
    }
  }
  testID="view"
/>
`;

exports[`withTwrn web should render non media queries styles along media queries styles 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff0000",
      "padding": 4,
    }
  }
  testID="view"
/>
`;

exports[`withTwrn web should respect queries order styles 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#00ff00",
      "padding": 4,
    }
  }
  testID="view"
/>
`;
